torch/_inductor/fx_passes/decompose_mm.py

```diff
@@ -0,0 +1,576 @@
+                        reduce_op=match.kwargs["reduce_op"],
+                        scatter_dim=0,
+                        group_size=match.kwargs["group_size"],
+                        group_name=match.kwargs["group_name"],
+                    )
+                )
+                continue
+
+    for reduce_scatter in reversed(reduce_scatters):
+        _process_matmul_reduce_scatter(
+            g, reduce_scatter, min_size_after_split, num_chunks
+        )
+
+    trace_structured(
+        "artifact",
+        metadata_fn=lambda: {
+            "name": "pass_split_mm_rs_graph_after",
+            "encoding": "string",
+        },
+        payload_fn=lambda: gm.print_readable(
+            print_output=False, include_stride=True, include_device=True
+        ),
+    )
+
+
+@dataclasses.dataclass
+class _PostMMOp:
+    n: torch.fx.Node
+
+    def apply(self, t, num_chunks):
+        return self.n.target(  # pyrefly: ignore[not-callable]
+            t, *self.n.args[1:], **self.n.kwargs
+        )
+
+
+@dataclasses.dataclass
+class _ViewOp(_PostMMOp):
+    chunk_dim: int
+
+    def apply(self, t, num_chunks):
+        shape = list(self.n.args[1])  # pyrefly: ignore[no-matching-overload]
+        if shape[self.chunk_dim] != -1:
+            shape[self.chunk_dim] //= num_chunks
+        return self.n.target(t, shape)  # pyrefly: ignore[not-callable]
+
+
+def _process_matmul_reduce_scatter(
+    g, reduce_scatter: _ReduceScatterMatch, min_size_after_split, num_chunks: list[int]
+) -> None:
+    (
+        input_node,
+        _reduce_scatter_node,
+        rs_wait_tensor_node,
+        reduce_op,
+        orig_scatter_dim,
+        group_size,
+        group_name,
+    ) = (
+        reduce_scatter.input_node,
+        reduce_scatter.reduce_scatter_node,
+        reduce_scatter.wait_tensor_node,
+        reduce_scatter.reduce_op,
+        reduce_scatter.scatter_dim,
+        reduce_scatter.group_size,
+        reduce_scatter.group_name,
+    )
+    if len(input_node.users) != 1:
+        return
+
+    # Find matmul node by walking backwards through the graph
+    mm_n, intermediate_nodes = _find_matmul_node(input_node)
+    if mm_n is None:
+        return
+
+    # Validate and extract operations between matmul and reduce-scatter
+    post_mm_ops, view3d_n = _extract_post_mm_ops(intermediate_nodes, orig_scatter_dim)
+    if post_mm_ops is None:
+        return
+
+    arg_a = mm_n.args[0]
+    arg_b = mm_n.args[1]
+    arg_a_t = arg_a.meta["val"]
+    arg_b_t = arg_b.meta["val"]
+
+    _fn_to_trace = _mm_split_cat_rs_fn
+
+    view3d_size_to_split = None
+    if orig_scatter_dim != 0 and view3d_n is not None:
+        _fn_to_trace = _mm_split_cat_rs_fn_view3d
+        # dim0 of 3d reshape
+        arg1 = view3d_n.args[1]
+        assert arg1 is not None
+        view3d_size_to_split = arg1[0]  # pyrefly: ignore[index-error]
+
+    size_before_split = 1
+    cont_check_t = arg_a_t
+    if orig_scatter_dim == 0:
```

Comment:
If we reduce_scatter on dimension 0 for `A @ B`, shouldn't we chunk A (by row) rather than B (by column) ?

Reply:
For example we have original mm like this:, world_size = 2, scatter_dim = 0
```
A0
B0
A1
B1
```
After reduce_scatter we want on rank 0:
```
reduce(A0)
reduce(B0)
```
If we split on dim = 0, then we have to prepare world_size of non contiguous inputs.
```
A0
A1
```
-> reduce_scatter 0

```
B0
B1
```
-> reduce_scatter 1

So between 
- cutting B by columns(dim=1)
-  cutting A into world_size * num_chunks  and then cat for A*, B*...

I picked cutting B where it is one chunk.

Reply:
I was thinking some optimal cases like, if we split A by row in this case with 2 chunks, we don't need a 'cat' in the end and we can replace reduce-scatter with reduce directly. Basically the chunk/split operation applied on A has already achieved the effect of 'scatter' part of 'reduce-scatter'. 

But I'm not sure if this exposes the same amount of overlapping btw comp/comm.